tokio-codec/src/lines_codec.rs

```diff
@@ -38,22 +107,72 @@ fn without_carriage_return(s: &[u8]) -> &[u8] {
 
 impl Decoder for LinesCodec {
     type Item = String;
+    // TODO: in the next breaking change, this should be changed to a custom
+    // error type that indicates the "max length exceeded" condition better.
     type Error = io::Error;
 
     fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<String>, io::Error> {
-        if let Some(newline_offset) =
-            buf[self.next_index..].iter().position(|b| *b == b'\n')
-        {
-            let newline_index = newline_offset + self.next_index;
-            let line = buf.split_to(newline_index + 1);
-            let line = &line[..line.len()-1];
-            let line = without_carriage_return(line);
-            let line = utf8(line)?;
-            self.next_index = 0;
-            Ok(Some(line.to_string()))
-        } else {
-            self.next_index = buf.len();
-            Ok(None)
+        loop {
+            // Determine how far into the buffer we'll search for a newline. If
+            // there's no max_length set, we'll read to the end of the buffer.
+            let read_to = if let Some(limit) = self.max_length {
+                // If a max length is set, read max_length bytes from the
+                // *current* offset into the buffer (which may not be 0).
+                let read_to = limit + self.next_index + 1;
+                // Clamp to buf.len(), so we don't go out of bounds.
+                cmp::min(read_to, buf.len())
+            } else {
+                buf.len()
+            };
+
+            let newline_offset = buf[self.next_index..read_to]
+                .iter()
+                .position(|b| *b == b'\n');
+
+            if self.is_discarding {
+                let discard_to = if let Some(offset) = newline_offset {
+                    // If we found a newline, discard up to that offset and
+                    // then stop discarding. On the next iteration, we'll try
+                    // to read a line normally.
+                    self.is_discarding = false;
+                    offset + self.next_index + 1
+               } else {
+                   // Otherwise, we didn't find a newline, so we'll discard
+                   // everything we read. On the next iteration, we'll continue
+                   // discarding up to max_len bytes unless we find a newline.
+                   read_to
+               };
+               buf.advance(discard_to);
+               self.next_index = 0;
+               continue;
```

Comment:
The continue is not needed here.

Reply:
Yeah, I added it after the fact because I thought it made it clearer when we would and would not loop --- I can remove it if you disagree.

Reply:
Removed in 0924eb29ffc922a6471582797ef977e2199feb00.